src/cm_editor.ml

```diff
@@ -0,0 +1,107 @@
+
+  val onDidChange : t -> Js.Unit.t Event.t
+
+  val create : uri:Uri.t -> t
+end = struct
+  include CustomDocument
+  module OnDidChange = Event.Make (Js.Unit)
+
+  include
+    [%js:
+    val onDidChange : t -> OnDidChange.t [@@js.get]
+
+    val create :
+      uri:Uri.t -> onDidChange:OnDidChange.t -> dispose:(unit -> unit) -> t
+      [@@js.builder]]
+
+  let content t =
+    let uri = uri t in
+    let file_path = Uri.fsPath uri in
+    let bin = Path.of_string "ocamlobjinfo" in
+    Cmd.(output (Spawn { bin; args = [ file_path ] }))
+
+  let create ~(uri : Uri.t) =
+    let module EventEmitter = EventEmitter.Make (Js.Unit) in
+    let onDidChange_event_emitter = EventEmitter.make () in
+    let onDidChange_event = EventEmitter.event onDidChange_event_emitter in
+    let watcher =
+      Workspace.createFileSystemWatcher
+        (`String (Uri.fsPath uri))
+        ~ignoreCreateEvents:true ~ignoreDeleteEvents:true ()
+    in
+    let disposable =
+      FileSystemWatcher.onDidChange watcher
+        ~listener:(fun _ -> EventEmitter.fire onDidChange_event_emitter ())
+        ()
+    in
+    create ~uri ~onDidChange:onDidChange_event ~dispose:(fun () ->
+        Disposable.dispose disposable)
+end
+
+let resolveCustomEditor _instance ~document ~webviewPanel ~token:_ :
+    unit Promise.t =
+  let open Promise.Syntax in
+  let update_content document =
+    let+ content = Cm_document.content document in
+    match content with
+    | Ok file_content ->
+      let webview = WebviewPanel.webview webviewPanel in
+      WebView.set_html webview (Printf.sprintf "<pre>%s</pre>" file_content)
+    | Error e ->
+      let uri = Cm_document.uri document in
+      let () = log "Error while trying to read content from file: %s" e in
+      let path = Uri.path uri in
+      let (_ : 'a option Promise.t) =
+        Window.showErrorMessage
+          ~message:
+            (Printf.sprintf "Error while trying to read content from %s."
+               (Node.Path.basename path))
+          ()
+      in
+      ()
+  in
+  let disposable =
+    Cm_document.onDidChange document
+      ~listener:(fun () ->
+        let (_ : unit Promise.t) = update_content document in
+        ())
+      ()
+  in
+  let (_ : Disposable.t) =
+    WebviewPanel.onDidDispose webviewPanel
+      ~listener:(fun () -> Disposable.dispose disposable)
+      ()
+  in
+  update_content document
+
+let openCustomDocument _instance ~(uri : Uri.t) ~openContext:_ ~token:_ :
+    Cm_document.t Promise.t =
+  let document = Cm_document.create ~uri in
+  Promise.resolve document
+
+let register (extension : ExtensionContext.t) (instance : Extension_instance.t)
+    =
+  let module CustomReadonlyEditorProvider =
+    CustomReadonlyEditorProvider.Make (Cm_document) in
+  let editorProvider =
+    CustomReadonlyEditorProvider.create
+      ~resolveCustomEditor:(resolveCustomEditor instance)
+      ~openCustomDocument:(openCustomDocument instance)
```

Comment:
am I missing something or `openCustomDocument` definition doesn't use `instance` so we could simply remove that arg?

Reply:
Oh yes, you're totally right. `resolveCustomEditor ` doesn't use it either. Removed in https://github.com/ocamllabs/vscode-ocaml-platform/pull/798/commits/6bbbdc8c156543a549b3162da0656d16a2bf62b0